Objects/unicodeobject.c

```diff
@@ -13302,6 +13393,29 @@ _PyUnicodeWriter_WriteSubstring(_PyUnicodeWriter *writer, PyObject *str,
     return 0;
 }
 
+
+int
+PyUnicodeWriter_WriteSubstring(PyUnicodeWriter *writer, PyObject *str,
+                               Py_ssize_t start, Py_ssize_t end)
+{
+    if (!PyUnicode_Check(str)) {
+        PyErr_Format(PyExc_TypeError, "expect str, not %T", str);
```

Comment:
Would not be better to raise SystemError to distinguish programming errors from from data driven errors? In all current use cases of `_PyUnicodeWriter_WriteSubstring()` it is impossible to pass wrong arguments, and if this happens, it is a programming error.

Reply:
TypeError is when the user pass the wrong type. Previously, we used SystemError more often. But it seems like the new trend (C API Working Group recommendations) is more to accept a generic `PyObject*` and raises TypeError if the function gets the wrong type.

Reply:
Recent example: PyLong_GetSign() raises TypeError, not SystemError.